### PR TITLE
Week9/calculator expression

### PR DESCRIPTION
**Pull request by koryun23**

### CODE REVIEW

File: src/main/com/aca/homework/week9/calculator/SimpleCalculator.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        if (!isValidExpression(expression)) {
+            System.out.println("error");
+            return "error";
+        }
+
+        int firstDigit = getDigitFromChar(expression.charAt(0));
+        int secondDigit = getDigitFromChar(expression.charAt(2));
+        char operation = expression.charAt(1);
+        switch (operation) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/SimpleCalculator.java:22]: the method is too big. Let's create an Operation enum class and implement calculate method for each enum value

Comment (gevorgyanarman) [reply src/main/com/aca/homework/week9/calculator/SimpleCalculator.java:22]: also create a static method in enum class that will take a char as an argument and will return enum value

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

File: src/main/com/aca/homework/week9/calculator/Operation.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    DIVISION {
+        @Override
+        public String performOperation(int num1, int num2) {
+            return String.valueOf((double) num1 / num2);
+        }
+    };
+
+    public abstract String performOperation(int num1, int num2);
+
+    public static Operation charToOperation(char ch) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/Operation.java:32]: let's name the method of(char c)

File: src/main/com/aca/homework/week9/calculator/SimpleCalculator.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        int firstDigit = getDigitFromChar(expression.charAt(0));
+        int secondDigit = getDigitFromChar(expression.charAt(2));
+        char operation = expression.charAt(1);
+        return Operation.charToOperation(operation).performOperation(firstDigit, secondDigit);
+    }
+
+    public boolean isValidExpression(String expression) {
+        return expression.length() == 3 && isDigit(expression.charAt(0)) && isDigit(expression.charAt(2)) && isDigit(expression.charAt(1));
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/SimpleCalculator.java:26]: expression.charAt(1) should be an operation, not a digit. I guess your tests are failing

File: src/main/com/aca/homework/week9/calculator/SimpleCalculator.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public boolean isOperation(char ch) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/SimpleCalculator.java:29]: you can move this in Operation class

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

File: src/main/com/aca/homework/week9/calculator/StringSupplier.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.aca.homework.week9.calculator;
+
+public interface StringSupplier {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/StringSupplier.java:3]: please create another implementation that ask an input from a user

File: src/main/com/aca/homework/week9/calculator/Operation.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
+package com.aca.homework.week9.calculator;
+
+public enum Operation {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/Operation.java:3]: Please create OperationTest class too and write unit tests for of(), isOperation() and performOperation() methods

File: src/main/com/aca/homework/week9/calculator/SimpleCalculator.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        char operation = expression.charAt(1);
+        return Operation.of(operation).performOperation(firstDigit, secondDigit);
+    }
+
+    public boolean isValidExpression(String expression) {
+        return expression.length() == 3 && isDigit(expression.charAt(0)) && isDigit(expression.charAt(2)) && Operation.isOperation(expression.charAt(1));
+    }
+
+
+    public boolean isDigit(char ch) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/SimpleCalculator.java:30]: please write unit tests for isDigit method too

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

File: src/main/com/aca/homework/week9/calculator/StringSupplier.java (hunk 5 above)
Comment (koryun23) [reply src/main/com/aca/homework/week9/calculator/StringSupplier.java:3]: Can't I just use anonymous class with the ```get``` method implemented like this: ```return new Scanner(System.in).nextLine()```?

**Comment by koryun23** [comment]
modified

File: src/main/com/aca/homework/week9/calculator/Operation.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    DIVISION {
+        @Override
+        public String performOperation(int num1, int num2) {
+            return String.valueOf((double) num1 / num2);
+        }
+    };
+
+    public abstract String performOperation(int num1, int num2);
+
+    public static Operation of(char ch) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/Operation.java:32]: let's use a for loop to return the operation for the given char

File: src/main/com/aca/homework/week9/calculator/Operation.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                return Operation.SUBTRACTION;
+            case '*':
+                return Operation.MULTIPLICATION;
+            case '/':
+                return Operation.DIVISION;
+            default:
+                return null;
+        }
+    }
+    public static boolean isOperation(char ch) {
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/Operation.java:46]: let's use a for loop to check

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

File: src/main/com/aca/homework/week9/calculator/Operation.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    },
+    DIVISION {
+        @Override
+        public String performOperation(int num1, int num2) {
+            return String.valueOf((double) num1 / num2);
+        }
+    };
+
+    private static final Map<Character, Operation> map = Map.of('+', Operation.ADDITION, '-', Operation.SUBTRACTION, '*', Operation.MULTIPLICATION, '/', Operation.DIVISION);
```
Comment (gevorgyanarman) [inline src/main/com/aca/homework/week9/calculator/Operation.java:32]: just pass a char as a constructor argument. there is no need to use a map

**Comment by koryun23** [comment]
done

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

File: src/main/com/aca/homework/week9/calculator/Operation.java (hunk 10 above)
Comment (koryun23) [reply src/main/com/aca/homework/week9/calculator/Operation.java:32]: don't I need a map for implementing the ``Operation of()`` and ``boolean isOperation()`` methods?

Comment (gevorgyanarman) [reply src/main/com/aca/homework/week9/calculator/Operation.java:32]: please pass + to constructor for ADDITION value,  to SUBTRACTION and so on. the using enum.values() filter and find the enum for the provided char

**Review by gevorgyanarman — CHANGES_REQUESTED** [review]

**Comment by koryun23** [comment]
modified

**Pull request merged**